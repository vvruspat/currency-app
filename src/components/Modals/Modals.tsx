import React, { useCallback, useState } from "react";
import {
  SelectModal,
  SelectModalProps,
} from "../uikit/Select/SelectModal/SelectModal";

export const ModalsContext = React.createContext<{
  setModal: (modal: ModalsProps | null) => void;
}>({
  setModal: (modal: ModalsProps | null) => {},
});

export enum MODALS {
  SELECT_MODAL = "select-modal",
  ERROR_MODAL = "error-modal",
  EXCHANGE_SUCCESS_MODAL = "exchange-success-modal",
}

// TODO: render and start modals from the component where it need

type ModalsProps = {
  modal: MODALS.SELECT_MODAL;
  props: SelectModalProps;
};
// | { modal: MODALS.ERROR_MODAL; props: /*ErrorModalProps*/ {} }
// | {
//     modal: MODALS.EXCHANGE_SUCCESS_MODAL;
//     props: /*ExchangeSuccessModalProps*/ {};
//   };

type ModalsRootProps = {};

export const ModalsRoot: React.FC<ModalsRootProps> = ({ children }) => {
  const [currentModal, setCurrentModal] = useState<ModalsProps | null>(null);

  const setModal = useCallback((modal: ModalsProps | null) => {
    setCurrentModal(modal);
  }, []);

  return (
    <ModalsContext.Provider value={{ setModal }}>
      {children}
      {/* <ErrorModal
        {...currentModalProps}
        show={currentModal === MODALS.ERROR_MODAL}
      /> */}
      <SelectModal
        {...currentModal?.props}
        show={currentModal?.modal === MODALS.SELECT_MODAL}
      />
      {/* <ExchangeSuccessModal
        {...currentModalProps}
        show={currentModal === MODALS.EXCHANGE_SUCCESS_MODAL}
      /> */}
    </ModalsContext.Provider>
  );
};
