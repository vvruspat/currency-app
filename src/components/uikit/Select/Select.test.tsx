import { fireEvent, render, screen } from "@testing-library/react";
import { ReactNode } from "react";
import { Caption, Select } from "..";
import { MODALS, ModalsRoot } from "../../Modals/Modals";

const props = {
  value: {
    key: "Item1",
    value: {
      icon: <div data-testid="test1-icon">Test1 icon element</div>,
      description: <Caption>Test1 Description</Caption>,
      content: "Test1 content",
    },
  },
  modalNav: MODALS.TEST_MODAL,
  options: [
    {
      key: "Item1",
      value: {
        icon: <div data-testid="test1-icon">Test1 icon element</div>,
        description: <Caption>Test1 Description</Caption>,
        content: "Test1 content",
      },
    },
    {
      key: "Item2",
      value: {
        icon: <div data-testid="test2-icon">Test2 icon element</div>,
        description: <Caption>Test1 Description</Caption>,
        content: "Test2 content",
      },
    },
    {
      key: "Item3",
      value: {
        icon: <div data-testid="test3-icon">Test3 icon element</div>,
        description: <Caption>Test3 Description</Caption>,
        content: "Test3 content",
      },
    },
    {
      key: "Item4",
      value: {
        icon: <div data-testid="test1-icon">Test4 icon element</div>,
        description: <Caption>Test4 Description</Caption>,
        content: "Test4 content",
      },
    },
  ],
};

test("renders select", async () => {
  const wrapper = ({ children }: { children: ReactNode }) => (
    <ModalsRoot>{children}</ModalsRoot>
  );

  render(<Select {...props} />, { wrapper });

  const select = screen.getByTestId("select-test");

  const optionElements = screen.getAllByText(/icon element/i);
  expect(optionElements[0]).toBeInTheDocument();
  expect(optionElements[3]).toBeInTheDocument();

  const buttonElement = screen.getByTestId("select-buton");
  expect(buttonElement).toBeInTheDocument();

  const modal = screen.getByTestId("select-modal");
  expect(modal).not.toHaveClass("modal-show");

  expect(select).toMatchSnapshot();

  fireEvent.click(buttonElement);

  expect(modal).toHaveClass("modal-show");

  const searchInputElement = screen.getByPlaceholderText("Search");
  expect(searchInputElement).toBeInTheDocument();

  expect(select).toMatchSnapshot();

  fireEvent.change(searchInputElement, { target: { value: "Item4" } });

  const searchInputElementUpdated = screen.getByDisplayValue("Item4");
  expect(searchInputElementUpdated).toBeInTheDocument();

  expect(select).toMatchSnapshot();
});
