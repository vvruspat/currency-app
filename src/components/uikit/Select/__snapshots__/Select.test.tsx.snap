// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders select 1`] = `
<div
  data-testid="select-test"
>
  <button
    class="icon-button icon-right"
    data-testid="select-buton"
  >
    <div
      class="icon-button-content"
    >
      <div
        class="icon-button-icon"
      >
        <svg>
          Down16Icon.svg
        </svg>
      </div>
      <div
        class="icon-button-text"
      >
        Test1 content
      </div>
    </div>
  </button>
  <div
    class="modal-overlay modal-overlay-hide"
  />
  <div
    class="modal"
    data-testid="select-modal"
  >
    <header
      class="modal-header"
    >
      <div
        class="modal-header-content"
      >
        <div
          class="caption"
        >
          <input
            class="input input-align-left select-modal-filter-input"
            data-testid="select-modal-filter-input"
            placeholder="Search"
            value=""
          />
        </div>
      </div>
      <button
        class="icon-button modal-header-close-button"
      >
        <div
          class="icon-button-content"
        >
          <div
            class="icon-button-icon"
          >
            <svg>
              Cross16Icon.svg
            </svg>
          </div>
          <div
            class="icon-button-text"
          >
            Close
          </div>
        </div>
      </button>
    </header>
    <section
      class="modal-content"
    >
      <div
        style="position: relative;"
      >
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test1-icon"
            >
              Test1 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test1 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test1 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test2-icon"
            >
              Test2 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test2 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test1 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test3-icon"
            >
              Test3 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test3 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test3 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test1-icon"
            >
              Test4 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test4 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test4 Description
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;

exports[`renders select 2`] = `
<div
  data-testid="select-test"
>
  <button
    class="icon-button icon-right"
    data-testid="select-buton"
  >
    <div
      class="icon-button-content"
    >
      <div
        class="icon-button-icon"
      >
        <svg>
          Down16Icon.svg
        </svg>
      </div>
      <div
        class="icon-button-text"
      >
        Test1 content
      </div>
    </div>
  </button>
  <div
    class="modal-overlay modal-overlay-show"
  />
  <div
    class="modal modal-show"
    data-testid="select-modal"
  >
    <header
      class="modal-header"
    >
      <div
        class="modal-header-content"
      >
        <div
          class="caption"
        >
          <input
            class="input input-align-left select-modal-filter-input"
            data-testid="select-modal-filter-input"
            placeholder="Search"
            value=""
          />
        </div>
      </div>
      <button
        class="icon-button modal-header-close-button"
      >
        <div
          class="icon-button-content"
        >
          <div
            class="icon-button-icon"
          >
            <svg>
              Cross16Icon.svg
            </svg>
          </div>
          <div
            class="icon-button-text"
          >
            Close
          </div>
        </div>
      </button>
    </header>
    <section
      class="modal-content"
    >
      <div
        style="position: relative;"
      >
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test1-icon"
            >
              Test1 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test1 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test1 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test2-icon"
            >
              Test2 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test2 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test1 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test3-icon"
            >
              Test3 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test3 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test3 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test1-icon"
            >
              Test4 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test4 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test4 Description
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;

exports[`renders select 3`] = `
<div
  data-testid="select-test"
>
  <button
    class="icon-button icon-right"
    data-testid="select-buton"
  >
    <div
      class="icon-button-content"
    >
      <div
        class="icon-button-icon"
      >
        <svg>
          Down16Icon.svg
        </svg>
      </div>
      <div
        class="icon-button-text"
      >
        Test1 content
      </div>
    </div>
  </button>
  <div
    class="modal-overlay modal-overlay-show"
  />
  <div
    class="modal modal-show"
    data-testid="select-modal"
  >
    <header
      class="modal-header"
    >
      <div
        class="modal-header-content"
      >
        <div
          class="caption"
        >
          <input
            class="input input-align-left select-modal-filter-input"
            data-testid="select-modal-filter-input"
            placeholder="Search"
            value="Item4"
          />
        </div>
      </div>
      <button
        class="icon-button modal-header-close-button"
      >
        <div
          class="icon-button-content"
        >
          <div
            class="icon-button-icon"
          >
            <svg>
              Cross16Icon.svg
            </svg>
          </div>
          <div
            class="icon-button-text"
          >
            Close
          </div>
        </div>
      </button>
    </header>
    <section
      class="modal-content"
    >
      <div
        style="position: relative;"
      >
        <div
          class="select-option"
          style="position: absolute; top: 0px; left: 0px; right: 0px; opacity: 1;"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test1-icon"
            >
              Test1 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test1 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test1 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
          style="position: absolute; top: 0px; left: 0px; right: 0px; opacity: 1;"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test2-icon"
            >
              Test2 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test2 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test1 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
          style="position: absolute; top: 0px; left: 0px; right: 0px; opacity: 1;"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test3-icon"
            >
              Test3 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test3 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test3 Description
              </div>
            </div>
          </div>
        </div>
        <div
          class="select-option"
          style="position: absolute; top: 0px; left: 0px; right: 0px; opacity: 1;"
        >
          <div
            class="select-option-icon"
          >
            <div
              data-testid="test1-icon"
            >
              Test4 icon element
            </div>
          </div>
          <div
            class="select-option-content"
          >
            <div
              class="select-option-content-data"
            >
              Test4 content
            </div>
            <div
              class="select-option-content-description"
            >
              <div
                class="caption"
              >
                Test4 Description
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
